### PR TITLE
Add ability to inject async client init params

### PR DESCRIPTION
**Pull request by odedlaz**

### CODE REVIEW

File: twistes/client.py [hunk 1]
```diff
@@ -14,19 +14,35 @@
...
         self._hostname, self._auth = self._es_parser.parse_host(hosts)
         self._timeout = timeout
-        self._async_http_client = async_http_client
+        self._async_http_client = async_http_client or treq
+        self._async_http_client_params = async_http_client_params or {}
         self.bulk_utils = BulkUtility(self)
 
+        if self._async_http_client == treq \
+                and 'pool' not in self._async_http_client_params:
+            self.inject_pool_to_treq(self._async_http_client_params)
```
Comment (avihad) [inline twistes/client.py:39]: add test for the injected params

File: twistes/client.py [hunk 2]
```diff
@@ -628,3 +645,22 @@ def _bulk_body(body):
...
+
+        def _check_fds(_):
+            fds = set(reactor.getReaders() + reactor.getReaders())
+            if not [fd for fd in fds if isinstance(fd, Client)]:
+                return
+
+            return deferLater(reactor, 0, _check_fds, None)
+
+        pool = self._async_http_client_params["pool"]
+        return pool.closeCachedConnections().addBoth(_check_fds)
```
Comment (avihad) [inline twistes/client.py:666]: add test that verify the method call

File: twistes/client.py (hunk 1 above)
Comment (odedlaz) [inline twistes/client.py:39]: done!

File: twistes/client.py (hunk 2 above)
Comment (odedlaz) [inline twistes/client.py:666]: done!

File: tests/test_elasticsearch.py [hunk 3]
```diff
@@ -39,27 +39,80 @@
 
 
 class TestElasticsearch(TestCase):
+
     def setUp(self):
         self.es = Elasticsearch(SOME_HOSTS_CONFIG, TIMEOUT, MagicMock())
 
+    def test_default_async_class(self):
+        es = Elasticsearch(SOME_HOSTS_CONFIG, TIMEOUT)
```
Comment (sxamit) [inline tests/test_elasticsearch.py:47]: import treq here instead of at the top of the file

**Comment by sxamit** [comment]
what's the trial maker?

File: tests/test_elasticsearch.py [hunk 4]
```diff
@@ -39,27 +39,80 @@
...
+        async_client = MagicMock()
+        async_client.request = MagicMock(return_value=ok_response)
+
+        async_client_params = {"test": "test"}
+        es = Elasticsearch(SOME_HOSTS_CONFIG,
+                           TIMEOUT,
+                           async_client,
+                           async_client_params)
+
+        result = yield es._perform_request(HttpMethod.GET, "/", None)
```
Comment (avihad) [inline tests/test_elasticsearch.py]: We don't want to test private methods, use wrapped calls like get or search

File: tests/test_elasticsearch.py [hunk 5]
```diff
@@ -39,27 +39,80 @@
...
+        auth = (SOME_USER, SOME_PASS)
+
+        expected_url = self._generate_url(SOME_HOST, SOME_PORT, None)
+        es._async_http_client.request.assert_called_once_with(HttpMethod.GET,
```
Comment (avihad) [inline tests/test_elasticsearch.py]: You can assert async_client.request instead

File: tests/test_elasticsearch.py (hunk 4 above)
Comment (odedlaz) [inline tests/test_elasticsearch.py]: ok

File: tests/test_elasticsearch.py (hunk 5 above)
Comment (odedlaz) [inline tests/test_elasticsearch.py]: done

File: tests/test_elasticsearch.py (hunk 3 above)
Comment (odedlaz) [inline tests/test_elasticsearch.py:47]: done

**Pull request merged**